src/libmongoc/src/mongoc/mongoc-collection.h

```diff
@@ -351,6 +351,30 @@ mongoc_collection_estimated_document_count (
    bson_t *reply,
    bson_error_t *error);
 
+
+
+
+typedef struct _mongoc_index_model_t mongoc_index_model_t;
+typedef struct _mongoc_index_options_t mongoc_index_options_t;
+
+MONGOC_EXPORT (bool)
+mongoc_collection_create_indexes (
+   mongoc_collection_t *collection,
+   const mongoc_index_model_t **indexes,
+   size_t num_indexes,
+   const bson_t *create_indexes_opts,
+   char **index_names, /* out parameter, populated with index names */
+   int *num_indexes_created, /* out parameter contained length of index_names */
+   bson_t *reply,
+   bson_error_t *error);
+
+MONGOC_EXPORT (mongoc_index_model_t *)
+mongoc_index_model_new (const bson_t *key, const bson_t *index_opts);
```

Comment:
I considered having `set_key` and `set_index_opts` functions but I like it a little more to take these at index_model creation time but I don't feel that strongly.

Reply:
To your comment, I definitely don't think we need extra helpers to set options. If anything, I'd argue for an even simpler API.

The index management spec doesn't seem to give a clear definition of IndexModel, so I imagine it's just inferred from the proposed method APIs and [language examples](https://github.com/mongodb/specifications/blob/master/source/index-management.rst#examples).

At the command level, the index model is really just a document where `key` and `name` are always defined (driver populates `name` if not provided by the user) and other options may follow in the same document. Does it make sense for the index model to be anything more than a `bson_t*` internally, even if you intend to abstract it for the user (i.e. define the struct as an opaque type but the `private.h` header defines it as a `bson_t*`)?

In the case where index model is just a `bson_t*` under the hood, you then wouldn't have to do anything special when crafting the `createIndexes` command for the singular or multi-index helper methods. And the pseudo-code for constructing a new index model would roughly be:

 * Allocate a new `bson_t*` for the index model
 * Set the `key` field from the explicit argument here
 * If `name` isn't already in the options, generate it from the key and assign it in the document
 * Copy in any other options (`copy_to_excluding` BSON helpers may work here) into the index model